### ISSUE TITLE
[HOLD for payment 2023-10-02] [HOLD for payment 2023-09-27] [Wave 6: Categories] CRITICAL: Allow Editing a Category on a Money Request

### ISSUE THREAD

**Issue by puneetlath** [Expensify/App#24464]
Tracking Issue: https://github.com/Expensify/Expensify/issues/282161

Design Doc: https://docs.google.com/document/d/1itXKWlhms5YHcF4-lujO6g8T5LPdLVD_7onjtZBsATQ/edit#

**HOLD ON https://github.com/Expensify/Expensify/issues/302031**

---

## Viewing Categories on Money Requests

For viewing a category on an already existing Money Request, if [`isSingleTransactionView`](https://github.com/Expensify/App/blob/a1e4e411cd89d3a2ea79b4d2232fdbe479e193d7/src/components/MoneyRequestHeader.js#L37) = true, we'll render a [`MenuItemWithTopDescription`](https://github.com/Expensify/App/blob/a87b802a09c48de7274855af56e27ab544463ef6/src/components/MenuItemWithTopDescription.js) in the [`MoneyRequestView`](https://docs.google.com/document/d/17jFcMsAotXvsMhXivoF3u7fyzeyXkKpRGXR88Ud0oWY/edit#bookmark=kix.gwxe6ynnvyp6). (This is not too different from what we do for the [`MoneyRequestConfirmationList`](https://docs.google.com/document/d/1itXKWlhms5YHcF4-lujO6g8T5LPdLVD_7onjtZBsATQ/edit#bookmark=id.13ms7czufn8) above when creating an expense).
- This menu item component will display the category tied to the expense.
- If there are no categories on the policy AND there is no category already on the expense, then the component will _not_ display.

If a request is paid, then users will _not_ be able to tap the request to edit the category, unless they are an admin of the policy the request is on.
- We can check if someone is an admin of a policy by calling [`PolicyUtils.isPolicyAdmin`](https://github.com/Expensify/App/blob/a602bc2a6f7b4f23f0b39d69b0209b47a95338d7/src/libs/PolicyUtils.js#L95-L101).

## Editing Categories on Money Requests
To edit, we will be building off the EditMoneyRequestPage added in [Wave 3: Manual Requests](https://docs.google.com/document/d/17jFcMsAotXvsMhXivoF3u7fyzeyXkKpRGXR88Ud0oWY/edit#bookmark=id.axyudzw5ui08). We'll add a new route to the `moneyRequestDetails` Form that will render our [`CategoryPicker`](https://docs.google.com/document/d/1itXKWlhms5YHcF4-lujO6g8T5LPdLVD_7onjtZBsATQ/edit#heading=h.4qev1cnytpps) described above.
```JS
// src/pages/iou/EditMoneyRequestPage.js

<Form
  formID='moneyRequestDetails'
  validate={this.validate}
  onSubmit={this.onSubmit}
>

    // ... Other routes and their edit paths

    ) ? route === 'category' (
      <CategoryPicker
        inputID="category"
        label="Category"
        placeholder={props.translate('common.category')}
      />
    )
</Form>
```

The component will work the same as it does in the Money Request creation flow, except this time the **`onSubmit`** prop function will call the `UpdateTransaction` API command. The action is described in [Wave 3: Manual Requests](https://docs.google.com/document/d/17jFcMsAotXvsMhXivoF3u7fyzeyXkKpRGXR88Ud0oWY/edit#bookmark=kix.hnekymqzjxyt). We will update it to support changing the category.

### A couple of other notes about editing categories:
- If you are offline, we will not actually call the `UpdateTransaction` API command but use the `optimisticData` of the `UpdateTransaction` `API.write` call to display your category in a "greyed out" state. When the update syncs with the server we'll "ink in" the change.
- If a category is disabled, but was already selected on a Money Request, when the user clicks the category to edit it, the disabled but selected category will show in the "Selected Category" section. Unselecting it will make it disappear since the user will be navigated back to the expense, saving an empty category on it.
    - Should they open the list again, because the category was disabled, they will not be able to see it to re-select it.
    - The same would apply if there were < 8 categories on the policy. We would just show the selected category directly in the list of categories itself.
- If categories are completely disabled, or there are no enabled categories on the policy, but there is a category on an expense…
    - When the user clicks the category to edit it, the lone category will be shown as the only category in the policy. 
    - This will use the "simplified" < 8 categories view.
    - Deselecting the category will again navigate the user back to the expense.
    - This time, because there are no categories on the policy and no category is selected, the Category menu item will disappear.

**Comment by rezkiy37** [comment]
Hi, I'm Michael (Mykhailo) from Callstack and I would like to work in this issue.

**Comment by waterim** [comment]
Hello, Im Artem from Callstack and I would like to help with this issue

**Comment by puneetlath** [comment]
This is still on hold.

**Comment by yuwenmemon** [comment]
Still on HOLD

**Comment by yuwenmemon** [comment]
Also, another development, because we want to start sending push notifications for category updates from our server before this goes live, we can no longer rely on the absence of categories to hide the selector from the `EditRequestPage`. We'll need to check that a user is on a beta in addition to that.

This means we'll need to add a canUseCategories to the [Permissions.js](https://github.com/Expensify/App/blob/50d3a09342660894d385966dc2f85e54b1853470/src/libs/Permissions.js) file, and check that before showing the selector or not. Does that make sense / sound good?

**Comment by yuwenmemon** [comment]
Removing HOLD!

**Comment by rezkiy37** [comment]
@puneetlath, I guess there is a minor mistake in the description of **Viewing Categories on Money Requests**. ~~**[MoneyRequestHeader](https://github.com/Expensify/App/blob/main/src/components/MoneyRequestHeader.js)**~~ **[MoneyRequestView](https://github.com/Expensify/App/blob/main/src/components/ReportActionItem/MoneyRequestView.js)**.

**Comment by yuwenmemon** [comment]
Yes, the MoneyRequestView is what we want to adjust. Some pieces got moved since we wrote that doc. I'll update the terms here in the description of the GH.

**Comment by rezkiy37** [comment]
Waiting on backend, because it is important to test and verify that the app handles everything properly.

**Comment by yuwenmemon** [comment]
Actually, @rezkiy37 we need to deploy the front-end portion first because of the way `EditMoneyRequest` works. Right now, if it doesn't see a category coming from the client it will clear the category entirely (cc @mountiny). Would you be able to build out the front-end portions - sending the category as a param to `EditMoneyRequest` into the ether - and then we can test and verify things work well from there? This is all under a beta so we should be safe to do so.

**Comment by rezkiy37** [comment]
> Actually, @rezkiy37 we need to deploy the front-end portion first because of the way `EditMoneyRequest` works. Right now, if it doesn't see a category coming from the client it will clear the category entirely (cc @mountiny). Would you be able to build out the front-end portions - sending the category as a param to `EditMoneyRequest` into the ether - and then we can test and verify things work well from there? This is all under a beta so we should be safe to do so.

Makes sense. I agree. I am preparing a PR today.

**Comment by rezkiy37** [comment]
Melvin, this issue is not completed yet. The main PR - https://github.com/Expensify/App/pull/27459 - still under review.

**Comment by puneetlath** [comment]
Looks like no C+ payment is required for this issue. So we can go ahead and close it out once it's on production.
